ddprof-ffi/src/tags.rs

```diff
@@ -0,0 +1,137 @@
+    let key = key.to_utf8_lossy().into_owned();
+    let value = value.to_utf8_lossy().into_owned();
+    match Tag::new(key, value) {
+        Ok(tag) => {
+            vec.push(tag);
+            PushTagResult::Ok
+        }
+        Err(err) => PushTagResult::Err(err.as_bytes().to_vec().into()),
+    }
+}
+
+#[repr(C)]
+pub struct ParseTagsResult {
+    tags: crate::Vec<Tag>,
+    error_message: Option<Box<crate::Vec<u8>>>,
+}
+
+#[must_use]
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_Vec_tag_parse(string: CharSlice) -> ParseTagsResult {
+    match unsafe { string.try_to_utf8() } {
+        Ok(string) => {
+            let (tags, error) = parse_tags(string);
+            ParseTagsResult {
+                tags: tags.into(),
+                error_message: error
+                    .map(|message| Box::new(crate::Vec::from(message.into_bytes()))),
+            }
+        }
+        Err(err) => ParseTagsResult {
+            tags: crate::Vec::default(),
+            error_message: Some(Box::new(crate::Vec::from(&err as &dyn Error))),
+        },
+    }
+}
+
+#[must_use]
+#[allow(clippy::ptr_arg)]
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_Vec_tag_clone(tags: &crate::Vec<Tag>) -> crate::Vec<Tag> {
+    tags.iter()
+        .map(|tag| tag.clone().into_owned())
+        .collect::<Vec<Tag>>()
+        .into()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::tags::*;
+
+    #[test]
+    fn empty_tag_name() {
+        unsafe {
+            let mut tags = ddprof_ffi_Vec_tag_new();
+            let result =
+                ddprof_ffi_Vec_tag_push(&mut tags, CharSlice::from(""), CharSlice::from("woof"));
+            assert!(!matches!(result, PushTagResult::Ok));
+        }
+    }
+
+    #[test]
+    fn test_lifetimes() {
+        let mut tags = ddprof_ffi_Vec_tag_new();
+        unsafe {
+            // make a string here so it has a scoped lifetime
+            let key = String::from("key1");
+            {
+                let value = String::from("value1");
+                let result = ddprof_ffi_Vec_tag_push(
+                    &mut tags,
+                    CharSlice::from(key.as_str()),
+                    CharSlice::from(value.as_str()),
+                );
+
+                assert!(matches!(result, PushTagResult::Ok));
+            }
+        }
+        let tag = tags.last().unwrap();
+        assert_eq!(tag.key(), "key1");
+        assert_eq!(tag.value(), "value1");
+    }
+
+    #[test]
+    fn test_dup() {
+        unsafe {
+            let mut tags = ddprof_ffi_Vec_tag_new();
+            let result = ddprof_ffi_Vec_tag_push(
+                &mut tags,
+                CharSlice::from("sound"),
+                CharSlice::from("woof"),
+            );
+            assert!(matches!(result, PushTagResult::Ok));
+
+            let mut cloned_tags = ddprof_ffi_Vec_tag_clone(&tags);
+            let result = ddprof_ffi_Vec_tag_push(&mut cloned_tags, "host".into(), "dog".into());
+            assert!(matches!(result, PushTagResult::Ok));
```

Comment:
Should we perhaps test here that `tags` was not modified, but `cloned_tags` was? It's slightly weird that this test would pass if I made `clone` a no-op ;)

Reply:
I'll improve it anyway, but Rust wouldn't let you make it a no-op :)